src/stable_marriage.rs

```diff
@@ -1,96 +1,234 @@
+    ) -> Option<&'a mut Woman<Self>> {
+        targets
+            .iter_mut()
+            .filter(|x| x.current_match.is_none())
+            .find(|x| self.is_perfect_match(item, &x.data))
     }
+}
 
-    (men, women)
+struct Man<P: Matcher + ?Sized> {
+    data: P::Item,
+    // Most preferred last
+    prefs: Vec<usize>,
 }
 
-fn extract_matching(v: Vec<Individual>) -> Vec<Option<usize>> {
-    v.into_iter().map(|x| x.cnt_match).collect()
+impl<P: Matcher + ?Sized> Man<P> {}
+
+pub struct Woman<P: Matcher + ?Sized> {
+    data: P::Target,
+    current_match: Option<Man<P>>,
+    current_is_perfect: bool,
 }
 
-fn saturate_matching(men: &mut Vec<Individual>, women: &mut Vec<Individual>) {
-    while let Some(i_man) = men
-        .iter()
-        .position(|m| m.cnt_match.is_none() && !m.prefs.is_empty())
+impl<P: Matcher + ?Sized> Woman<P> {
+    fn prefers_to_current<Q>(&self, matcher: &Q, item: &P::Item) -> bool
+    where
+        Q: Matcher<Item = P::Target, Target = P::Item>,
     {
-        while let Some(i_woman) = men[i_man].pop_preferred_if_unmatched() {
-            let woman = &mut women[i_woman];
-
-            if let Some(i_man_in_w_prefs) = woman.get_pref_idx(i_man) {
-                if let Some(i_otherman) = woman.cnt_match {
-                    men[i_otherman].cnt_match = None;
-                }
-                men[i_man].cnt_match = Some(i_woman);
-                woman.cnt_match = Some(i_man);
-                woman.prefs.truncate(i_man_in_w_prefs);
-            }
+        if self.current_is_perfect || !matcher.is_admissible(&self.data, item) {
+            return false;
+        }
+        if let Some(ref cnt_man) = self.current_match {
+            matcher.cmp_3way(&self.data, &cnt_man.data, item) == Ordering::Greater
+        } else {
+            true
         }
     }
+
+    fn replace_match(&mut self, man: Man<P>) -> Option<Man<P>> {
+        let mut old_match = Some(man);
+        std::mem::swap(&mut self.current_match, &mut old_match);
+        old_match
+    }
 }
 
 // Computes a stable matching between two lists of individuals.
 // See https://en.wikipedia.org/wiki/Stable_marriage_problem
 // This implements an extended version of the Gale-Shapley algorithm that allows for some
 // individuals to not rank every individual from the other list, in which case those two
 // individuals will never be matched together. In particular, the lists need not be the same size.
-//
-// @arg men: A list of preference rankings (most preferred first) of indices in the list `women`
-// @arg women: A list of preference rankings (most preferred first) of indices in the list `men`
-pub fn stable_marriage(
-    men: Vec<Vec<usize>>,
-    women: Vec<Vec<usize>>,
-) -> (Vec<Option<usize>>, Vec<Option<usize>>) {
-    let (mut men, mut women) = init_matching(men, women);
-
-    saturate_matching(&mut men, &mut women);
-
-    (extract_matching(men), extract_matching(women))
+// This algorithm favors men.
```

Comment:
Hmm… I'm not 100% sure, but it more or less looks like this is no longer a stable matching, but a greedy matching from the man's perspective. Or are the two actually identical? (thinking esp. of a man who'd be evicted by another man later in the list, who'd never have a chance to meet the woman after his first in his list)

Reply:
The two are identical, and this algorithm follows the standard Gale-Shapley algorithm you can find anywhere (modulo handling of allowed_divergence and having unmatched individuals).

Reply:
The `find_perfect_match` function only looks at women that do not already have a match, so this shouldn't happen. I added a comment